pkg/kubelet/dockertools/docker_cache.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockertools
+
+import (
+	"sync"
+	"time"
+)
+
+type DockerCache interface {
+	KubeletRunningDockerContainers() (DockerContainers, error)
+}
+
+func NewDockerCache(client DockerInterface) (DockerCache, error) {
+	containers, err := GetKubeletDockerContainers(client, false)
+	if err != nil {
+		return nil, err
+	}
+	return &dockerCache{
+		client:        client,
+		cacheTime:     time.Now(),
+		containers:    containers,
+		updatingCache: false,
+	}, nil
+}
+
+// dockerCache is a default implementation of DockerCache interface
+type dockerCache struct {
+	// The underlying docker client used to update the cache.
+	client DockerInterface
+
+	// Mutex protecting all of the following fields.
+	lock sync.Mutex
+	// Last time when cache was updated.
+	cacheTime time.Time
+	// The content of the cache.
+	containers DockerContainers
+	// Whether the background thread updating the cache is running.
+	updatingCache bool
+	// Last time when the cache was examined.
+	lastReqTime time.Time
+}
+
+func (d *dockerCache) KubeletRunningDockerContainers() (DockerContainers, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if time.Since(d.cacheTime) > 2*time.Second {
+		containers, err := GetKubeletDockerContainers(d.client, false)
+		if err != nil {
+			return containers, err
+		}
+		d.containers = containers
+		d.cacheTime = time.Now()
+	}
+	d.lastReqTime = time.Now()
```

Comment:
I see - it wasn't exactly 100ms - it was 100ms + "time to get running containers from Docker". But yeah - I just change it to be 2 seconds since the last request.